docs/architecture/adr-008-ics20-implementation.md

```diff
@@ -0,0 +1,435 @@
+The  implementation is broken down into traits which should be implemented by the ICS20 module
+context, it also defines some primitives that would help in building a module compliant with the ICS20 spec.
+
+Decided it's better for the ICS20 context to be completely independent of the ibc core context traits, that way it can be 
+fully implemented as a standalone module in any framework.
+
+Coupling the ICS20 Context with the ibc Core traits will not allow the existence of the ICS20 module as a standalone library in some frameworks
+It should be upto the module implementer to use the provided helper functions and ICS20 primitives correctly
+
+```rust
+    define_error! {
+    #[derive(Debug, PartialEq, Eq)]
+    Error {
+        InvalidDenomTrace
+            | _ | { "Denom trace is not valid" },
+
+        SendDisabled
+            | _ | { "Sending tokens is disabled" },
+        ReceiveDisabled
+            | _ | { "Receiving tokens is disabled" },
+        }
+    }
+    #[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord, Serialize, Deserialize)]
+    #[serde(untagged)]
+    pub struct Denom(String);
+    
+    impl AsRef<&str> for Denom {
+        fn as_ref(&self) -> &str{
+            self.0.as_str()
+        }
+    }
+
+    impl FromStr for Denom{
+        fn from_str(val: &str) -> Self {
+            Self(val.to_string())
+        }
+    }
+
+    /// Coin defines a token with a denomination and an amount.
+    #[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord, Serialize, Deserialize)]
+    pub struct Coin {
+        /// Denomination
+        pub denom: DenomTrace,
+
+        /// Amount
+        pub amount: U256,
+    }
+
+    pub struct TracePrefix {
+        pub port_id: PortId,
+        pub channel_id: ChannelId
+    }
+
+    /// This struct and it's implementations should help identifying denomination traces
+    pub struct DenomTrace {
+        /// Full denomination path
+        pub trace_path: Vec<TracePrefix>,
+        /// Base denimination
+        pub base_denom: Denom
+    }
+
+    impl DenomTrace {
+        /// Returns the full denom path
+        fn get_full_denom_path(&self) -> String {
+            self.trace_path.into_iter().fold(String::new(), |acc, val| {
+                let next = Self::get_denom_prefix(&val.port_id, &val.channel_id);
+                format!("{}{}", acc, next)
+            })
+        }
+        /// IBCDenom a coin denomination for an ICS20 fungible token in the format
+        /// 'ibc/trace_path/base_denom'. If the trace is empty, it will return the base denomination.
+        fn ibc_denom(&self) -> String {
+            format!("ibc/{}", self.get_full_denom_path())
+        }
+        /// Returns the prefix for this trace
+        fn get_prefix(&self) -> String {
+            DenomTrace::get_denom_prefix(self.trace_path[0].port_id, self.trace_path[0].channel_id)
+        }
+
+        /// Returns the prefix for this trace
+        pub fn has_prefix(denom: String, prefix: String) -> bool {
```

Comment:
`denom` and `prefix` should be `&str`. Same goes for all other functions in the ADR that take a `String` as input